lib/IRGen/GenMeta.cpp

```diff
@@ -5532,6 +5518,60 @@ namespace {
       B.addNullPointer(IGM.Int8PtrTy);
     }
   };
+
+  /// A builder for ForeignReferenceTypeMetadata.
+  class ForeignReferenceTypeMetadataBuilder :
+      public ForeignMetadataBuilderBase<ForeignReferenceTypeMetadataBuilder,
+                                        ForeignClassMetadataBuilderBase<ForeignReferenceTypeMetadataBuilder>> {
+  public:
+    ForeignReferenceTypeMetadataBuilder(IRGenModule &IGM, ClassDecl *target,
+                                        ConstantStructBuilder &B)
+        : ForeignMetadataBuilderBase(IGM, target, B) {
+      assert(getTargetType()->isForeignReferenceType() &&
+             "foreign reference type metadata build must be used on foreign reference types.");
+
+      if (IGM.getOptions().LazyInitializeClassMetadata)
+        CanBeConstant = false;
+    }
+
+    void emitInitializeMetadata(IRGenFunction &IGF, llvm::Value *metadata,
+                                MetadataDependencyCollector *collector) {
+      llvm_unreachable("Not implemented for foreign reference types.");
+    }
+
+    // Visitor methods.
+
+    void addValueWitnessTable() {
+      auto type = getTargetType()->getCanonicalType();
+      B.add(irgen::emitValueWitnessTable(IGM, type, false, false).getValue());
+    }
+
+    void addMetadataFlags() {
+      B.addInt(IGM.MetadataKindTy, (unsigned) MetadataKind::ForeignReferenceType);
+    }
+
+    void addNominalTypeDescriptor() {
+      auto descriptor =
+          ClassContextDescriptorBuilder(this->IGM, Target, RequireMetadata).emit();
+      B.addSignedPointer(descriptor,
+                         IGM.getOptions().PointerAuth.TypeDescriptors,
+                         PointerAuthEntity::Special::TypeDescriptor);
+    }
+
+    void addSuperclass() {
+      // Always leave the superclass pointer unfilled.  We'll have to
+      // unique it during initialization anyway, so we might as well spare
+      // ourselves the load-time work.
+      B.addNullPointer(IGM.TypeMetadataPtrTy);
```

Comment:
I don't see this superclass field in the metadata class.

In general, if there isn't any commonality of layout in the metadata classes, I'm not sure we should be using related metadata builder classes; it just ends up obscuring how the code in GenMeta matches the structs in Metadata.h.

Reply:
This is using `ForeignMetadataBuilderBase` which seemed generic enough. I'll just make this into a no-op + assert. Let me know if you want me to add another builder, though.

Reply:
Looking at this again, I see that I am actually using `ForeignClassMetadataBuilderBase`. I added my own builder for FRTs and updated it to use that. Thanks for pointing this out.